File: leetcode/20. Valid Parentheses.rs
use std::collections::HashMap;

impl Solution {
    pub fn is_valid(s: String) -> bool {
        let mut left = Vec::new();
        let m = HashMap::from([
            ('}', '{'),
            (']', '['),
            (')', '('),
        ]);
        
        for c in s.chars() {
            if c == '{' || c == '[' || c == '(' {
                left.push(c);
            } else {
                if let Some(&lc) = m.get(&c) {
                    if let Some(expect) = left.pop() {
                        if lc != expect {
                            return false; 
                        }
                    } else {
                        return false;
                    }
                }
            }
        }
        left.len() == 0
    }
}

File: leetcode/string/20. Valid Parentheses.py
class Solution:
    def isValid(self, s: str) -> bool:
        stack = []
        mapping = {
            '}': '{',
            ')': '(',
            ']': '['
        }
        for c in s:
            if c in {'(', '[', '{'}:
                stack.append(c)
            else:
                if len(stack) == 0 or stack[-1] != mapping[c]:
                    return False
                stack.pop()
        return len(stack) == 0